clang/lib/CodeGen/Targets/AMDGPU.cpp

```diff
@@ -537,7 +537,8 @@ AMDGPUTargetCodeGenInfo::getLLVMSyncScopeID(const LangOptions &LangOpts,
     break;
   }
 
-  if (Ordering != llvm::AtomicOrdering::SequentiallyConsistent) {
+  if (LangOpts.OpenCL &&
```

Comment:
I am deeply opposed to making any __builtin_amdgcn intrinsic behavior language dependent. This is a target builtin with target defined behavior that should behave consistently regardless of language mode. I'd rather just take an OpenCL performance regression than this

Reply:
If OpenCL wants this funny behavior, they should need to use a different builtin. That said, is OpenCL even using the generic builtin?

Reply:
This is not an AMDGPU builitin, the AMDGPU builtin takes a string that corresponds directly to the syncscope string. This is for the 'generic' versions which was reimplemented like three times for OpenCL, HIP, and the GNU / scoped versions. I think that a better solution here would be to only set it if it uses the OpenCL flavored scope arguments, then HIP and C/C++ get the non `one-as` implementation, as those targets don't explicitly use the same AS knowledge.

Reply:
My point is 1 builtin : 1 behavior, universally. No modality.

Reply:
The amdgcn fence builtin also looks like it already takes optional arguments to append MMRAs. If OpenCL is actually using this builtin anywhere (which I doubt it is), it can easily migrate to __builtin_amdgcn_fence

Reply:
That's the current behavior now that I updated it.

Reply:
I'm sure it's using it somewhere since it's the OpenCL spec I think? This currently miscompiles on C/C++ w/ OpenMP so I think it should be fixed. Possible performance regressions but I don't think it's a valid optimization on non-OpenCL targets since we don't explicitly expose the address spaces as a part of the language.

Reply:
If we don't have tests running __scoped_atomic_* through OpenCL, we should have them